### PR TITLE
Challenge AI as color from handicap position

### PR DESCRIPTION
**Pull request by ddugovic**
Use Fairy-Stockfish to ease CI setup and Linux development setup
Fix recent CI regressions
Also fix challenge timeout creating many games
Also fix engine not receiving depth etc. parameters causing engine to lose on time
Challenge AI as color from handicap position

### CODE REVIEW

File: config.yml.default [hunk 1]
```diff
@@ -22,11 +22,11 @@ engine:                        # Engine Settings.
...
-#   USI_Hash: 256              # Max memory (in megabytes) the engine can allocate.
-#   go_commands:               # Additional options to pass to the USI go command.
-#     nodes: 1                 # Search so many nodes only.
-#     depth: 5                 # Search depth ply only.
-#     movetime: 1000           # Integer. Search exactly movetime milliseconds.
+    Hash: 256                  # Max memory (in megabytes) the engine can allocate.
+  go_commands:                 # Additional options to pass to the USI go command.
+    nodes: 50000               # Search so many nodes only.
+    depth: 21                  # Search depth ply only.
+    movetime: 500              # Integer. Search exactly movetime milliseconds.
```
Comment (TheYoBots) [inline config.yml.default:29]: Thanks for your pull request!
Just a few changes...
Here, could you change `Hash` to `USI_Hash`? It works even for fairy stockfish when I run `./fairy-stockfish usi`.

Also the `go_commands` could be commented, so that there is no specified commands pushed by default.

File: config.yml.default [hunk 2]
```diff
@@ -2,8 +2,8 @@ token: "xxxxxxxxxxxxxxxx"      # Lishogi OAuth Token.
 url: "https://lishogi.org/"    # Lishogi base URL.
 
 engine:                        # Engine Settings.
-  dir: "./engines/"            # Directory containing the engine. This can be an absolute path or one relative to Lishogi-Bot/.
-  name: "engine_name"          # Binary name of the engine to use. Make sure the engine you use is running under the USI protocol.
+  dir: "/usr/games/"           # Directory containing the engine. This can be an absolute path or one relative to Lishogi-Bot/.
+  name: "fairy-stockfish"      # Binary name of the engine to use. Make sure the engine you use is running under the USI protocol.
```
Comment (TheYoBots) [inline config.yml.default:6]: Can this be undone and instead done in the `test.py` file?
```py
    CONFIG["engine"]["dir"] = "./usr/games/"
    CONFIG["engine"]["name"] = "fairy-stockfish"
```

Also I honestly prefer testing with Yaneuraou, but since this does reduce some unneeded code I suppose it is better.

File: requirements.txt [hunk 3]
```diff
@@ -3,3 +3,5 @@ PyYAML==6.0
 requests==2.28.2
 backoff==2.2.1
 rich==13.3.3
+pytest==7.2.2
+pytest-timeout==2.1.0
```
Comment (TheYoBots) [inline requirements.txt:7]: Are these changes required? It seemed to run fine when these dependencies were directly downloaded in the github actions file right.

File: lishogi.py [hunk 4]
```diff
@@ -126,6 +130,6 @@ def set_user_agent(self, username):
         self.header.update({"User-Agent": f"Lishogi-Bot/{self.version} user:{username}"})
         self.session.headers.update(self.header)
 
-    def challenge_ai(self):
-        data = {"level": 1, "clock.limit": 60, "clock.increment": 0, "clock.byoyomi": 2, "clock.periods": 1}
-        return self.api_post(ENDPOINTS["challenge_ai"], data=data)
\ No newline at end of file
+    def challenge_ai(self, color="random", sfen="lnsgkgsnl/1r5b1/ppppppppp/9/9/9/PPPPPPPPP/1B5R1/LNSGKGSNL b - 1"):
+        challenge = {"level": 1, "color": color, "sfen": sfen, "clock.limit": 60, "clock.increment": 0, "clock.byoyomi": 2, "clock.periods": 1}
+        return self.api_post(ENDPOINTS["challenge_ai"], data=challenge, timeout=30)
```
Comment (TheYoBots) [inline lishogi.py:135]: I don't think these changes are required either. `"random"` is the default colour and `"lnsgkgsnl/1r5b1/ppppppppp/9/9/9/PPPPPPPPP/1B5R1/LNSGKGSNL b - 1"` is the default `sfen`. They do not need to be set. Adding `timeout=30` is fine though here.

File: engine_ctrl/usi.py [hunk 5]
```diff
@@ -8,10 +8,10 @@
 
 
 class Engine:
-    def __init__(self, variants, command, cwd=None):
+    def __init__(self, command, cwd=None):
         self.info = {}
         self.id = {}
-        self.variants = variants
+        #self.variants = variants
```
Comment (TheYoBots) [inline engine_ctrl/usi.py:14]: Could you remove this tooo please?

File: engine_wrapper.py [hunk 6]
```diff
@@ -92,8 +94,8 @@ def search(self, sfen, moves, btime=None, wtime=None, binc=None, winc=None, byo=
         return best_move, ponder_move
 
     def print_stats(self, stats=None):
-        for line in self.get_stats(stats=stats):
-            logger.info(f"{line}")
+        lines = self.get_stats(stats=stats)
+        logger.info(', '.join(lines))
```
Comment (TheYoBots) [inline engine_wrapper.py:98]: What does this change do differently compared to its previous behavior?

File: engine_ctrl/usi.py (hunk 5 above)
Comment (ddugovic) [reply engine_ctrl/usi.py:14]: Ah, sorry about this one; your change on line 11 (adding argument "variants") broke this bot in ways I can't understand yet.

File: engine_wrapper.py (hunk 6 above)
Comment (ddugovic) [reply engine_wrapper.py:98]: This consolidates score, depth, nodes, and nps onto the same line so the log doesn't flood as much:
`INFO     score: {'cp': 330}, depth: 6, nodes: 1258, nps: 179714           engine_wrapper.py:98`

File: lishogi.py (hunk 4 above)
Comment (ddugovic) [reply lishogi.py:135]: OK, I'll keep these in my fork and remove them from the PR.  Long-term I'd like to add xboard support and variant support, so having robust options I can easy change for testing purposes will be useful... but you're right that you don't need these options right now.

File: requirements.txt (hunk 3 above)
Comment (ddugovic) [reply requirements.txt:7]: New developers wanting to test need instructions about how to set up test dependencies.  I suppose these could go in a readme instead.

File: config.yml.default (hunk 1 above)
Comment (ddugovic) [reply config.yml.default:29]: Regarding Hash versus USI_Hash, you're the maintainer and I can revert this for compatibility, although the UCI/USI specification and Fairy-Stockfish (at least on my PC) expect Hash instead.

Commenting go_commands would seem to negate potential value of having a test framework, but yeah having these uncommented would just confuse users so I'll comment them out again.

File: config.yml.default (hunk 2 above)
Comment (ddugovic) [reply config.yml.default:6]: Yes, I'll undo this and use `tests.py` .

Sorry, I was having a headache trying to debug Yaneuraou on Linux in a way that wouldn't unzip itself and pollute my workspace... I think for new developers, Fairy-SF is easiest to work with (on any OS) and has excellent minishogi support.

File: engine_ctrl/usi.py (hunk 5 above)
Comment (TheYoBots) [reply engine_ctrl/usi.py:14]: python-shogi doesn't support minishogi, so I doubt it would support the promotion rule of minishogi. Currently, for variants, so bot depends on the engine and doesn't check any rules which is why illegal moves might be possible with certain engines.

File: engine_wrapper.py (hunk 6 above)
Comment (TheYoBots) [reply engine_wrapper.py:98]: I'd prefer if you'd revert this. It's much neater.

File: requirements.txt (hunk 3 above)
Comment (ddugovic) [reply requirements.txt:7]: @TheYoBots Sorry, I should ask: do you want me to revert this?  This seems like useful information for new developers & having an agreed-upon test library version could be useful.

Comment (TheYoBots) [reply requirements.txt:7]:
```txt
python-shogi==1.0.16
PyYAML==6.0
requests==2.28.2
backoff==2.2.1
rich==13.3.3

# Requirements for tests
pytest==7.2.2
pytest-timeout==2.1.0
```
A comment before the test dependencies would be better then.

File: config.yml.default [hunk 7]
```diff
@@ -23,10 +23,10 @@ engine:                        # Engine Settings.
...
     Threads: 1                 # Max CPU threads the engine can use.
 #   USI_Hash: 256              # Max memory (in megabytes) the engine can allocate.
-# go_commands:                 # Additional options to pass to the USI go command.
-#   nodes: 1                   # Search so many nodes only.
-#   depth: 5                   # Search depth ply only.
-#   movetime: 1000             # Integer. Search exactly movetime milliseconds.
+#   go_commands:               # Additional options to pass to the USI go command.
+#     nodes: 1                 # Search so many nodes only.
+#     depth: 5                 # Search depth ply only.
+#     movetime: 1000           # Integer. Search exactly movetime milliseconds.
```
Comment (TheYoBots) [inline config.yml.default]: As per your changes to the `engine_wrapper.py` file, shouldn't you keep the whitespace changes?

Comment (ddugovic) [reply config.yml.default]: Oops... yes, you are correct.

File: engine_ctrl/usi.py (hunk 5 above)
Comment (ddugovic) [reply engine_ctrl/usi.py:14]: Ah, sorry now I remembered my other concern... if multiple opponents challenge an engine to multiple variants, and it's playing multiple games at the same time, how does that work?

I guess each variant could have a separate config. file or something like that, but that seems confusing.

Comment (TheYoBots) [reply engine_ctrl/usi.py:14]: Do you mean different engines for different variants?

File: engine_wrapper.py [hunk 8]
```diff
@@ -95,6 +97,7 @@ def print_stats(self, stats=None):
         for line in self.get_stats(stats=stats):
             logger.info(f"{line}")
 
+
```
Comment (TheYoBots) [inline engine_wrapper.py:100]: Just a minor change, could please remove this line? Most functions here have only a 1 line break.

File: lishogi.py [hunk 9]
```diff
@@ -127,5 +131,5 @@ def set_user_agent(self, username):
         self.session.headers.update(self.header)
 
     def challenge_ai(self):
-        data = {"level": 1, "clock.limit": 60, "clock.increment": 0, "clock.byoyomi": 2, "clock.periods": 1}
-        return self.api_post(ENDPOINTS["challenge_ai"], data=data)
\ No newline at end of file
+        challenge = {"level": 1, "clock.limit": 60, "clock.increment": 0, "clock.byoyomi": 2, "clock.periods": 1}
+        return self.api_post(ENDPOINTS["challenge_ai"], data=challenge, timeout=30)
```
Comment (TheYoBots) [inline lishogi.py:135]: Also could you increase the time here? The engine might be stronger than level 1, but might lose on time. Perhaps the byoyomi could be increased?

*Based on the test results in your repository (The one in my repository tests something else)*

File: engine_ctrl/usi.py (hunk 5 above)
Comment (ddugovic) [reply engine_ctrl/usi.py:14]: Thinking more about it... I suppose my question is: what purpose does this line of code serve?

If two opponents challenge a bot to two different variants, this function wouldn't be where the `new` and `variant` and `position ... moves` commands would be run anyway.  The game variant can differ there is a different game to consider.

I guess when constructing an engine it could be possible to check, "Does this engine actually support the variants listed in the config file?" but that isn't what this code currently does.

File: lishogi.py (hunk 9 above)
Comment (ddugovic) [reply lishogi.py:135]: I can however this timeout isn't the reason my repo is failing; I believe that's partially caused by a Fairy-SF time management issue... which I can use go commands in `tests.py` (rather than in the config file) to work around.

This timeout has to do with: the bot challenges AI level 1 to a game, but doesn't get a response until 5-10 seconds later after AI level 1 has made a move.  Setting `timeout=30` (instead of the default 2 seconds) prevents creation of additional games while waiting on AI level 1 to make its first move.

I discovered this pattern after I added the ability to challenge as sente/gote (instead of random).

File: engine_ctrl/usi.py (hunk 5 above)
Comment (TheYoBots) [reply engine_ctrl/usi.py:14]: If the variant isn't supported by the engine, then it would probably play an illegal move and give this warning message:
https://github.com/TheYoBots/Lishogi-Bot/blob/bdaad6701f715b9e9b9f039c74bbe193d91be996/engine_ctrl/usi.py#L102

Comment (TheYoBots) [reply engine_ctrl/usi.py:14]: I think this pull request is ready to be merged after this commented line is removed right?

Comment (ddugovic) [reply engine_ctrl/usi.py]: That sounds good to me; my final test is running at https://lishogi.org/@/FrogfuciusBot/tv
